### PR TITLE
`<vector>`: More scope guards

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Towards #2307. Possibly closes it?

There're still two occurrences of try-catch-reraise in `_Insert_counted_range`, which don't seem able to be replaced by scope guards if we want to keep the current behavior. It seems that no more `_RERAISE;` in other headers can be replaced with scope guards.

Drive-by: strengthening exception specification for
- `allocator<T>::deallocate`,
- `allocator_traits<allocator<T>>::deallocate`, and
- `vector<T, A>::_Change_array`.

### CODE REVIEW

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks, this is a very clean transformation! :magic_wand: I pushed a conflict-free merge with `main` and reordered one guard member.

File: stl/inc/vector [hunk 1]
```diff
@@ -592,6 +592,60 @@ private:
...
+    struct _NODISCARD _Vaporization_guard { // vaporize the detached piece
+        vector* _Target;
+        pointer _Vaporized_first;
+        pointer _Vaporized_last;
+        pointer _Destroyed_first;
+
+        _Vaporization_guard& operator=(const _Vaporization_guard&) = delete;
+        _Vaporization_guard& operator=(_Vaporization_guard&&)      = delete;
+
+        ~_Vaporization_guard() noexcept {
```
[the remark this answers is not in the document]
Comment (frederick-vs-ja) [reply stl/inc/vector:637]: We may need to add `_CONSTEXPR26` once WG21-P3068 gets adopted for C++26... but perhaps we should add it when that happens.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks again for improving the STL's bestest data structure! :100: :star: :green_heart:

**Pull request merged**